model/map/src/main/java/org/keycloak/models/map/role/MapRoleProvider.java

```diff
@@ -0,0 +1,361 @@
+        if (first != null && first >= 0) {
+            s = s.skip(first);
+        }
+        if (max != null && max >= 0) {
+            s = s.limit(max);
+        }
+        return s;
+    }
+
+    @Override
+    public Stream<RoleModel> getRealmRolesStream(RealmModel realm) {
+        return getNotRemovedUpdatedRolesStream()
+                .filter(entityRealmFilter(realm))
+                .filter(this::isRealmRole)
+                .sorted(COMPARE_BY_NAME)
+                .map(entityToAdapterFunc(realm));
+    }
+
+    private boolean isRealmRole(MapRoleEntity role) {
+        return ! role.isClientRole();
+    }
+
+    @Override
+    public RoleModel addClientRole(ClientModel client, String id, String name) {
+        if (getClientRole(client, name) != null) {
+            throw new ModelDuplicateException();
+        }
+
+        final UUID entityId = id == null ? UUID.randomUUID() : UUID.fromString(id);
+
+        LOG.tracef("addClientRole(%s, %s, %s)%s", client.getClientId(), id, name, getShortStackTrace());
+
+        MapRoleEntity entity = new MapRoleEntity(entityId, client.getRealm().getId());
+        entity.setName(name);
+        entity.setClientRole(true);
+        entity.setClientId(client.getId());
+        if (tx.get(entity.getId(), roleStore::get) != null) {
+            throw new ModelDuplicateException("Role exists: " + id);
+        }
+        tx.putIfAbsent(entity.getId(), entity);
+        return entityToAdapterFunc(client.getRealm()).apply(entity);
+    }
+
+    @Override
+    public Stream<RoleModel> getClientRolesStream(ClientModel client, Integer first, Integer max) {
+        Stream<RoleModel> s = getClientRolesStream(client);
+        if (first != null && first > 0) {
+            s = s.skip(first);
+        }
+        if (max != null && max >= 0) {
+            s = s.limit(max);
+        }
+        return s;
+    }
+
+    @Override
+    public Stream<RoleModel> getClientRolesStream(ClientModel client) {
+        return getNotRemovedUpdatedRolesStream()
+                .filter(entityRealmFilter(client.getRealm()))
+                .filter(entityClientFilter(client))
+                .sorted(COMPARE_BY_NAME)
+                .map(entityToAdapterFunc(client.getRealm()));
+    }
+    @Override
+    public boolean removeRole(RoleModel role) {
+        LOG.tracef("removeRole(%s(%s))%s", role.getName(), role.getId(), getShortStackTrace());
+
+        RealmModel realm = role.isClientRole() ? ((ClientModel)role.getContainer()).getRealm() : (RealmModel)role.getContainer();
+
+        session.users().preRemove(realm, role);
+
+        RoleContainerModel container = role.getContainer();
+        if (container.getDefaultRolesStream().anyMatch(r -> Objects.equals(r, role.getName()))) {
+            container.removeDefaultRoles(role.getName());
+        }
+
+        //remove role from realm-roles composites
+        try (Stream<RoleModel> baseStream = getRealmRolesStream(realm).filter(RoleModel::isComposite)) {
```

Comment:
Should we create `RoleModel` for each entity? We can work with entities and then turn to model only those that need to be updated (so that we add entities to transaction). Or we can maybe get rid of models here completely and work with entities only. Because if I see correctly, the MapRoleModel just call an operation on the entity. WDYT? @hmlnarik Maybe you could have also something to say to this.

I am asking also because I am doing something similar for users, so if it is incorrect I need to make few changes.

Reply:
I've changed it according to your suggestion. Thank you. The code is a little bit longer and maybe slightly less readable, but as you said by this approach we got rid of sorting and turning all entities to model. I'd like to also hear @hmlnarik's opinion on this.

Reply:
+1 to @mhajas's suggestion. Thanks for updating this!